### ISSUE TITLE
`unnecessary_overrides` doesn't work for Flutter `State` class

### ISSUE THREAD

**Issue by ookami-kb** [dart-lang/linter#4359]
The lint rule works for a plain Dart class, but doesn't show warning for `State` class:

<img width="1920" alt="image" src="https://github.com/dart-lang/linter/assets/853356/e22ad989-fea7-4817-91e6-702aac9248a8">

The example can be checked in dartpad: https://dartpad.dev/?id=ad789dfba46401b322dc6a7f6dd580bf

Flutter 3.10.0 Dart SDK 3.0.0

**Comment by srawlins** [comment]
Thanks @ookami-kb !

It appears to have something to do with the `@protected` annotation [on `State.initState`](https://api.flutter.dev/flutter/widgets/State/initState.html). Here's a minimal repro:

```dart
import 'package:meta/meta.dart';

class Foo {
  @protected
  void initState() {}
}

class Bar extends Foo {
  @override
  void initState() {
    super.initState();
  }
}
```

**Comment by ookami-kb** [comment]
Ah, that's interesting. Indeed, if I add `@protected` annotation to the overridden method, the lint will warn me about unnecessary override:

```dart
class Foo {
  @protected
  void initState() {}
}

class Bar extends Foo {
  @override
  @protected
  void initState() {
    super.initState();
  }
}

```

That makes sense since I'm actually exposing the method by making it public, but in practice, I doubt that the majority of developers actually add this annotation to overridden methods.

I wonder what would be the correct way to handle this situation in general.

**Comment by bwilkerson** [comment]
The lint rule decides that an override is necessary if either
- there's more than just an invocation of the overridden behavior,
- the types of the parameters or the return type is different,
- the overridden method has a doc comment, or
- the overridden method has a different set of annotations than the overridden method.

**Comment by srawlins** [comment]
Sounds like we can close this.
